Sources/Algorithms/Keyed.swift

```diff
@@ -0,0 +1,86 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+public struct KeysAreNotUnique<Key, Element>: Error {
+  public let key: Key
+  public let previousElement: Element
+  public let conflictedElement: Element
+
+  public init(key: Key, previousElement: Element, conflictedElement: Element) {
```

Comment:
I might be missing something really obvious, but… why is an explicit initialiser necessary?  Doesn't Swift implicitly generate a memberwise initialiser?

Reply:
The implicit generated initialiser is `internal` but this one is `public`. Also, you probably want to mark the init `@inlinable` because it is generic and otherwise will not spezialise outside of this module.

Reply:
I wasn't sure about `@inlinable`.

> otherwise will not spezialise outside of this module.

Since this is kind of an edge case, is the perf improvement worth it relative to the code size cost of specializing it? What's the SOP for deciding on that?

Reply:
> Since this is kind of an edge case, is the perf improvement worth it relative to the code size cost of specializing it? 

IIUC `@inlinable` will just make the body of function available for specialization outsize the module, but at the end is the optimizer which will decide based on a cost model if is going to inline or not. e.g. if user compiles with `-Osize` it may not inline that call at all even though is marked as inlinable. For given a hint that we always want to inline, we can use `@inline(__always)` together with `@inlinable`. 

> What's the SOP for deciding on that?

With that said, I think is always worth to add `@inlinable`. Specially in this case as code size impact would be not significant. But I also see why it wouldn't worth it